add json guide

tinnus-napbus opened the pull request:
- explains $json type
- explains relevant libraries
- explains how you'd approach the issue in the context of a gall agent
- provides example of encoding/decoding a structure to $json
- provides example of a json conversions with a mark file

note this doesn't backtick every single instance of everything for readability purposes. I have backticked things when it seemed appropriate, typically when the thing was first introduced in the relevant section

note this links out to the zuse json documentation in #1111 and the mark file documentation in #769 so depends on those (if you don't want dead links anyway)

resolves #1093

tinnus-napbus committed:
change diagram location to media.urbit.org

tinnus-napbus committed:
json guide: various phrasing/grammar fixes

tinnus-napbus committed:
json guide: more phrasing/style tweaks

drbeefsupreme commented on content/docs/hoon/guides/json-guide.md:
```diff
@@ -0,0 +1,363 @@
+
+In brief, a mark file contains a `door` with three arms. The door's sample type is the type of the data in question - in our case the `$user` structure. The `+grab` arm contains methods for converting _to_ our mark, and the `+grow` arm contains methods for converting _from_ our mark. The `+noun` arms are mandatory, and then we've added `+json` arms which respectively call the `+from-js` and `+to- ...
+
+From this mark file, Clay can build mark conversion `gate`s between the `%json` mark and our `%user` mark, allowing the conversion of `$json` data to a `$user` structure and vice versa.
```
```suggestion
From this mark file, Clay can build mark conversion gates between the `%json` mark and our `%user` mark, allowing the conversion of `$json` data to a `$user` structure and vice versa.
```

drbeefsupreme commented on content/docs/hoon/guides/json-guide.md:
````diff
@@ -0,0 +1,363 @@
+]
+```
+
+We'll also build our library so we can use its types from the Dojo:
````
```suggestion
We'll also build our library so we can use its types from the dojo:
```

drbeefsupreme commented on content/docs/hoon/guides/json-guide.md:
````diff
@@ -0,0 +1,363 @@
+> =json-to-user .^($-(json user:user-lib) %cf /===/json/user)
+```
+
+Let's try convert our `$json` to a `$user` structure with our new mark conversion gate:
````
```suggestion
Let's try converting our `$json` to a `$user` structure with our new mark conversion gate:
```

drbeefsupreme reviewed (changes requested):
LGTM. I think it would also be good to mention this guide at the top of `/hoon/guides/parsing.md`, like "if you're looking for a guide to parse JSON specifically, see here"

tinnus-napbus committed:
json guide: minor style fix - remove backticks from gate

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
json guide: uncapitalise dojo

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
json guide: minor grammar change

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
add note & link in parsing guide

tinnus-napbus committed:
json guide: minor grammar fix

basilesportif reviewed (changes requested):
OK, couple things left to do:

- note that your `from-js` gate is a parser, but also that the gates `ot` uses are JSON parsers as well. It's important to convey that recursive nature.
- decoding JSON is the thing that gives people the most trouble. So there should be more about patterns with `ot`/`of`. Link to or copy relevant source from the codebase for examples?
- should have explanation/examples for decoding optional fields
- cover `pairs` and the simple encoding primitives more

tinnus-napbus committed:
merge upstream resolve conflict in index

tinnus-napbus committed:
Merge branch 'urbit:master' into json-guide

tinnus-napbus committed:
more +dejs examples & use +pairs in +enjs example

tinnus-napbus commented:
ok @timlucmiptev I added a new section that covers `ou` and `of`, and I changed the enjs example to use pairs and added a note about frond. I also added a note emphasising the recursive nesting of dejs functions

tinnus-napbus committed:
json guide: fix typo

basilesportif commented:
That works for me.

tinnus-napbus committed:
json guide: minor link text fix

liam-fitzgerald commented on content/docs/hoon/guides/json-guide.md:
```diff
@@ -0,0 +1,453 @@
+  =,  dejs:format
+  |=  jon=json
+  ^-  user
+  %.  jon
```
This is not strictly necessary, cause these runes are effectively a no-op. I'd scrap these three lines and replace it with  `^-  $-(json user)` cast, to make it clear that the `dejs` reparsers are higher order functions

liam-fitzgerald commented on content/docs/hoon/guides/json-guide.md:
```diff
@@ -0,0 +1,453 @@
+
+Usually (though not in all cases) these mark conversions will be performed implicitly by Gall or Eyre and you'd not deal with the mark conversion gates directly, but it's still informative to see them work explicitly.
+
+## Further reading
```
Worth linking directly to `scot` imo, given its prevalence in JSON conversion.

liam-fitzgerald commented on content/docs/hoon/guides/json-guide.md:
```diff
@@ -0,0 +1,453 @@
+
+There are dozens of different functions in [`+dejs:format`](/docs/hoon/reference/zuse/2d_6) that will cover a great many use cases. If there isn't a `+dejs` function for a particular case, you can also just write a custom function - it just has to take `$json`. Note there's also the [`+dejs-soft:format`](/docs/hoon/reference/zuse/2d_7) functions - these are similar to `+dejs` functions except the ...
+
+## More `+dejs`
```
Nothing on `+su`? Imo one of the most important decoders, because there's a lot of pressure on JSON objects to be somewhat stringly-typed (keys must be strings, strings don't break pointer equality unexpectedly).

liam-fitzgerald reviewed (commented):
Looks good! Some minor nits, but looks terrific broadly. Thank you for this

tinnus-napbus committed:
json guide: change dejs example gate to 569XZims(json user) format

tinnus-napbus committed:
json guide: add strings guide link for scot usage

tinnus-napbus committed:
json guide: add +su section

tinnus-napbus commented:
@liam-fitzgerald cool I've addressed your comments and added a +su section

liam-fitzgerald reviewed (approved):
LGTM!
